_develop/keypress-on-entry.md

```diff
@@ -0,0 +1,114 @@
+
+order: 120
+---
+
+## Respond To Specific Keypress In Entry
+---
+
+In the traditional sense, GUI programs have used callbacks to customize actions for widgets. Fyne does not expose inserting closure callbacks inside widgets, but it does not need to. The Go language is plenty extensible to make this work.
+
+Simply use Type Embedding and extend the widget, along with adding a callback field for the specific action.
+
+First create a new type struct, we will call it `EnterEntry`. Then embed `widget.Entry` and insert a callback field called `OnEnter`.
+
+```go
+type EnterEntry struct {
+	widget.Entry
+
+	OnEnter func()
+}
+```
+
+As mentioned in [Extending existing widgets](https://fyne.io/develop/extending-widgets.html), we follow good practice  and create a constructor function and extend the `BaseWidget`.
+
+```go
+func newEnterEntry() *EnterEntry {
+	entry := &EnterEntry{}
+	entry.ExtendBaseWidget(entry)
+	return entry
+}
+```
+
+Then override the `KeyDown()` function that's part of `widget.Entry` by adding the following block of code into our source file. Inside this function, we will use a conditional to check if the `key.Name` value matches the `Return` key but using the `fyne.KeyReturn` variable (we could just as well have checked that key.Name == `"Return"`). This can easily be extended to check for other keys in the future if necessary.
+
+```go
+func (entry *EnterEntry) KeyDown(key *fyne.KeyEvent) {
+	switch key.Name {
+	case fyne.KeyReturn:
+		entry.OnEnter()
+	}
+}
+```
+
+Now things get interesting, we simply call our callback `OnEnter`. It doesn't do anything yet, but that's because we need to tell it  what to do using the `OnEnter` callback.
+
+Once we constuct our `EnterEntry`, we assign an anonymous function to `OnEnter`. It's important to do it this way because anonymous functions (or closures) can capture their surrounding environment of variables. This offers more flexibility to interact with more than one widget, such as passing text to a label and insert in a VBox to simulate chat history.
+
+```go
+messages := widget.NewVBox()
+
+entry := newEnterEntry()
+entry.OnEnter =	func() {
+	message := entry.Text
+	messages.Append(widget.NewLabel(message))
+	entry.SetText("")
+}
+```
+
+With this, `OnEnter()` now performs a custom action and will be actuated when the `KeyDown()` event occurs in our GUI program.
+
+In the end, the resulting code should look something like this:
+
+```go
+package main
+
+import (
+	"fyne.io/fyne"
+	"fyne.io/fyne/app"
+	"fyne.io/fyne/layout"
+	"fyne.io/fyne/widget"
+)
+
+type EnterEntry struct {
+	widget.Entry
+
+	OnEnter func()
+}
+
+func newEnterEntry() *EnterEntry {
+	entry := &EnterEntry{}
+	entry.ExtendBaseWidget(entry)
+	return entry
+}
+
+func (entry *EnterEntry) KeyDown(key *fyne.KeyEvent) {
+	switch key.Name {
+	case fyne.KeyReturn:
+		entry.OnEnter()
+	}
```

Comment:
I think it's important to pass KeyEvent to parent entry's `KeyDown` method when extending widgets (like calling `entry.Entry.KeyDown(key)` at the end of the method). if you don't do it then some things like selection via shift and keyboard arrows won't work. Or, if it was the intent, it's important to note this behavior in the tutorial.

Reply:
Agreed.